### PR TITLE
Constraints

### LINKED ISSUE #145: Mapper currently does existential quanification on unbound ports. Should be universal

### LINKED ISSUE #150: Support for constraining ADT types in mapper
Similar to #130 

Now that you can specify a set of values that constant bitvector fields can take during the solving process  (#144), it would be useful if this were extended to ADT types as well. Also, it would be useful if there were some mechanism for constraining multiple signals at once.
Ideally, I would like to be able to constrain the ALU instruction, LUT inputs, Cond instruction, and Config address/data all at the same time. 

Note that the instruction to the PE is a Product type with the ALU, LUT, and Cond instructions as fields. It would be useful if there were an easy way to constrain each of these separately, for example specifying ALU = (Add or Sub), LUT = 0, Cond = 0, instead of constraining the entire Inst signal to some concatenation of their bitvector representation.

### LINKED ISSUE #160: [Feature Request] Constrain rewrite rule generator values by path instead of type
Previous API:

`arch_mapper = ArchMapper(arch_fc, constrain_constant_bv : tp.Option[None, tuple])`

Proposed API:

```
constraints = {
  path1: (BV4(0),BV4(1), BV4(2)),
  path2: Bit(0)
  path3: (MyEnum.A, MyEnum.B)
  ...
}
arch_mapper = ArchMapper(arch_fc, path_constraints=constraints)
```

Any unspecified (unconstrained) path will:
-Existentially quantified if it is a Const type. 
-Universally quantified otherwise

### PR DESCRIPTION
**Pull request by rdaly525**
Fixes #160
Fixes #150 
Fixes #145 

See
https://github.com/cdonovick/peak/blob/475eaef4d2e817fb872c8c02045c9edea59d9f40/peak/mapper/mapper.py#L355
For description of how constraints modify the SMT formula

Does not address constraining Sum/Tagged Union choices.

### CODE REVIEW

**Comment by rdaly525** [comment]
also addresses #150

File: tests/test_aadt_mapping.py [hunk 1]
```diff
@@ -18,6 +19,9 @@ def test_automapper():
     expect_found = ('Add', 'Sub', 'And', 'Nand', 'Or', 'Nor', 'Not', 'Neg')
     expect_not_found = ('Mul', 'Shftr', 'Shftl')
     for ir_name, ir_fc in IR.instructions.items():
+        if ir_name in ('Not', 'Neg'):
+            #Not implemented yet
```
Comment (rdaly525) [inline tests/test_aadt_mapping.py:23]: This needs #164 in order to resolve

File: examples/sum_pe/isa.py [hunk 2]
```diff
@@ -1,18 +1,26 @@
...
-        class Opcode(Enum):
-            A  = 1
-            B  = 2
-        offset = Word
-    return Inst, Operand0T, Operand1T, Word, Tuple[Word, Bit]
+    # TODO there is a bug when trying to use Product here
+    #class BitOp(Product):
+    #    op = Op
+    #    neg = Bit
+
```
Comment (cdonovick) [inline examples/sum_pe/isa.py:19]: Remove commented code.  If there is a TODO document in as issue.

File: tests/test_aadt_mapping.py (hunk 1 above)
Comment (rdaly525) [reply tests/test_aadt_mapping.py]: Fixed

File: examples/sum_pe/isa.py (hunk 2 above)
Comment (rdaly525) [reply examples/sum_pe/isa.py]: Fixed

**Pull request merged**